Author: Tanay Yadav

Roll No: AI20BTECH11026

Assignment 2

Creating the .CSV files from the given datasets:  

The datasets were imported and the spaces between the pixel values were replaced with commas using the following code:  
\# *first get all lines from file*  
with open('file.txt', 'r') as f:  
&nbsp;   &nbsp;   lines = f.readlines()  

\# *remove spaces*   
&nbsp;   &nbsp;   lines = \[line.replace(' ', ',') for line in lines\]    
    
\# *finally, write lines in the file*    
with open('file.txt', 'w') as f:   
&nbsp;   &nbsp;   f.writelines(lines)  

Then finally the extension of the .txt file was changed to .csv

# Q5

In [101]:
#importing the required libraries

import numpy as np
import pandas as pd
from sklearn import svm
from sklearn.metrics import accuracy_score
import csv

In [102]:
# creating the dataframes and processing the data for fitting

train_data = pd.read_csv('gisette_train.csv', header = None)
test_data = pd.read_csv('gisette_valid.csv', header = None)
train_data = train_data.drop(columns = 5000)
test_data = test_data.drop(columns = 5000)

In [103]:
# displaying the training dataframe

train_data

0     1     2     3     4     5     6     7     8     9     ...  4990  \
0      550     0   495     0     0     0     0   976     0     0  ...     0   
1        0     0     0     0     0     0     0   976     0     0  ...     0   
2        0     0     0     0     0     0     0     0     0     0  ...     0   
3        0     0   742     0     0     0     0   684     0   956  ...     0   
4        0     0     0     0     0     0     0   608     0   979  ...   991   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
5995     0     0     0     0     0     0     0     0     0     0  ...   991   
5996     0     0     0     0     0     0     0     0     0     0  ...     0   
5997     0     0     0     0     0   758     0     0     0   522  ...     0   
5998     0     0     0     0     0     0     0     0     0     0  ...     0   
5999     0     0   991     0     0     0     0     0     0     0  ...     0   

      4991  4992  4993  4994  4995  4996  4997  4998  4999  
0        0     0   991   991     0     0     0     0   983  
1      475     0   991     0     0   991     0     0     0  
2        0     0     0     0     0     0     0     0     0  
3        0     0     0     0     0   674     0     0   838  
4        0     0   828     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
5995     0     0     0     0     0   783     0     0     0  
5996     0     0     0     0     0   921     0   886     0  
5997   901     0     0     0     0   980     0     0     0  
5998     0     0     0   690     0     0     0     0     0  
5999     0     0   288     0     0   972     0     0   991  

[6000 rows x 5000 columns]

In [104]:
# displaying the testing dataframe

test_data

0     1     2     3     4     5     6     7     8     9     ...  4990  \
0     688     0     0     0     0     0     0   952     0   870  ...     0   
1     778   758     0     0     0     0     0   708     0   991  ...     0   
2     469     0   816     0     0     0     0     0     0     0  ...     0   
3       0     0     0     0     0   571   991   983     0   983  ...     0   
4       0     0     0     0     0     0     0   949     0   991  ...     0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
995     0   511     0     0     0     0     0     0     0     0  ...   991   
996     0     0   599     0     0     0     0   910     0     0  ...     0   
997     0     0     0     0     0     0     0     0     0     0  ...     0   
998     0     0     0     0     0     0     0   968     0     0  ...     0   
999     0     0     0     0     0     0     0     0     0     0  ...     0   

     4991  4992  4993  4994  4995  4996  4997  4998  4999  
0       0     0     0     0     0   494     0   769     0  
1       0     0   770     0     0     0     0     0     0  
2       0     0   467     0     0     0   417     0     0  
3       0     0     0     0     0     0     0     0     0  
4       0     0     0     0     0     0     0   976     0  
..    ...   ...   ...   ...   ...   ...   ...   ...   ...  
995     0     0   852     0     0     0     0     0     0  
996     0     0     0     0     0     0     0     0   707  
997     0     0     0     0     0     0     0     0   742  
998     0     0   773     0     0     0     0     0     0  
999     0     0     0     0     0   952     0     0     0  

[1000 rows x 5000 columns]

In [105]:
# creating the arrays for training and testing labels using loadtxt function from numpy

with open('gisette_train_labels.csv') as f:
    train_label = np.loadtxt(f, delimiter = ',')
    
with open('gisette_valid_labels.csv') as f:
    test_label = np.loadtxt(f, delimiter = ',')

In [106]:
# verifying the shape of the array containing training labels

train_label.shape

(6000,)

In [107]:
# verifying the shape of the array containing testing labels

test_label.shape

(1000,)

# Q5(a)

In [108]:
# creating the classifier clf using the sklearn.svm library
# linear kernel 

clf = svm.SVC(kernel = 'linear')

In [109]:
# fitting the classifier using the training dataset

clf.fit(train_data, train_label)

SVC(kernel='linear')

In [110]:
# predicting the labels on the training dataset

y_train = clf.predict(train_data)

#calculating the accuracy of the predictions over the training dataset

train_accu = accuracy_score(train_label, y_train)
print('Training error =', 1 - train_accu)

Training error = 0.0


In [111]:
# predicting the labels over the testing dataset

y_test = clf.predict(test_data)

#calculating the accuracy of the predictions over the testing dataset

test_accu = accuracy_score(test_label, y_test)
print('Testing error =', 1 - test_accu)

Testing error = 0.02400000000000002


In [112]:
# finding the number of support vectors for both the classes

vectors = clf.n_support_
print('Number of support vectors =', vectors)

Number of support vectors = [542 542]


# Q5(b)

Gaussian Kernel with $\gamma = 0.001$

In [113]:
# creating the classifier clf using the sklearn.svm library
# rbf kernel

clf_rbf = svm.SVC(kernel = 'rbf', gamma = 0.0001)

# fitting the classifier using the training dataset
clf_rbf.fit(train_data, train_label)

# predicting the labels on the training dataset
y_rbf_train = clf_rbf.predict(train_data)

#calculating the accuracy of the predictions over the training dataset
rbf_train_accu = accuracy_score(train_label, y_rbf_train)

# predicting the labels over the testing dataset
y_rbf_test = clf_rbf.predict(test_data)

#calculating the accuracy of the predictions over the testing dataset
rbf_test_accu = accuracy_score(test_label, y_rbf_test)

# finding the number of support vectors for both the classes
rbf_vectors = clf_rbf.n_support_

print('For RBF Kernel:')
print('Training error =', 1 - rbf_train_accu)
print('Testing error =', 1 - rbf_test_accu)
print('Number of support vectors =', rbf_vectors)

For RBF Kernel:
Training error = 0.0
Testing error = 0.5
Number of support vectors = [3000 3000]


Polynomial Kernel with $C = 1$ and $Q = 2$ i.e. $(1 + x^Tx)^2$

In [114]:
# creating the classifier clf using the sklearn.svm library
# polynomial kernel

clf_poly = svm.SVC(kernel = 'poly', coef0 = 1, degree = 2)

# fitting the classifier using the training dataset
clf_poly.fit(train_data, train_label)

# predicting the labels on the training dataset
y_poly_train = clf_poly.predict(train_data)

#calculating the accuracy of the predictions over the training dataset
poly_train_accu = accuracy_score(train_label, y_poly_train)

# predicting the labels over the testing dataset
y_poly_test = clf_poly.predict(test_data)

#calculating the accuracy of the predictions over the testing dataset
poly_test_accu = accuracy_score(test_label, y_poly_test)

# finding the number of support vectors for both the classes
poly_vectors = clf_poly.n_support_

print('For Polynomial Kernel:')
print('Training error =', 1 - poly_train_accu)
print('Testing error =', 1 - poly_test_accu)
print('Number of support vectors =', poly_vectors)

For Polynomial Kernel:
Training error = 0.0004999999999999449
Testing error = 0.020000000000000018
Number of support vectors = [641 691]


Observations:  
#### For *Linear* Kernel:  
Training error $= 0.0 $   
Testing error $= 0.02400000000000002$  
Number of support vectors for both the classes $= 542$    

#### For *RBF* Kernel with $\gamma = 0.0001$:  
Training error $= 0.0$  
Testing error $= 0.5$  
Number of support vectors for both the classes $= 3000$  

#### For *Polynomial* Kernel with $C = 1$ and $Q = 2$:    
Training error $= 0.0004999999999999449$  
Testing error $= 0.020000000000000018$  
Number of support vectors for the class $'4'$ is $641$ and $'9'$ is $691$.

### Hence, the *RBF* Kernel with $\gamma = 0.0001$ yields the lowest Training Error of $0.0$